core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java

```diff
@@ -270,7 +272,20 @@ public CompletionStage<Configuration> getOrCreateCache(String cacheName, String
    }
 
    CompletionStage<Configuration> createCache(String cacheName, String template, Configuration configuration, EnumSet<CacheContainerAdmin.AdminFlag> flags) {
-      return createCacheInternal(cacheName, template, configuration, flags).thenApply((v) -> configuration);
+      CompletionStage<Configuration> cs = createCacheInternal(cacheName, template, configuration, flags)
+            .thenApply((v) -> configuration);
+      if (!CompletionStages.isCompletedSuccessfully(cs)) {
+         outstandingCreate.compute(cacheName, (ignore, existing) -> {
+            if (existing == null) return cs.thenRun(() -> outstandingCreate.remove(cacheName));
+
+            CompletableFuture<Void> o = existing.toCompletableFuture();
+            return cs.whenComplete((res, t) -> {
+               if (t != null) o.completeExceptionally(t);
+               else o.complete(null);
+            }).thenRun(() -> outstandingCreate.remove(cacheName));
```

Comment:
As we only call `outstandingCreate.remove(cacheName)` in `thenRun`, won't this mean that the `outstandingCreate` entry  will never be removed if the original stage completes exceptionally?